### PR TITLE
Games: Add Spider

### PR DESCRIPTION
**Pull request by jamierocks**
The bulk of the game, and the changes necessary to LibCards are laid out enough that some initial review can be done - and would be appreciated :) I'm particularly interested in how others think LibCards could be altered to support other patience games (e.g. Spider :p).

Screenshot_20210617_120436

### CODE REVIEW

**Comment by gunnarbeutner** [comment]
Very nice! One thing I've noticed is that sometimes the game won't let me drag&drop a card until after I've moved the target card:


https://user-images.githubusercontent.com/388571/122484745-8d449d80-cfd5-11eb-9fcc-712a09655e56.mp4

**Comment by gunnarbeutner** [comment]
Also, I guess this shouldn't happen (middle stack). Unfortunately I don't know what exactly I did:

Screenshot from 2021-06-18 02-02-03

**Comment by jamierocks** [comment]
Glad to see you're enjoying it :)

With regards to the video, I had that once but hadn't been able to reproduce it since - good to know that it's still around, will have to poke at that tommorow.

I have a suspicion that edge-case with a full set not completing is perhaps the result of a draw completing the set? Looking at the code quickly, that appears to me as the most obvious omission (again, will address in the morning).

File: Userland/Games/Spider/Game.cpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+                if (top_card.is_upside_down()) {
+                    if (top_card.rect().contains(click_location)) {
+                        top_card.set_upside_down(false);
+                        start_timer_if_necessary();
+                        update(top_card.rect());
+                    }
+                } else if (m_focused_cards.is_empty()) {
+                    to_check.add_all_grabbed_cards(click_location, m_focused_cards, Cards::CardStack::Same);
+                    m_mouse_down_location = click_location;
+                    to_check.set_focused(true);
```
Comment (gunnarbeutner) [inline Userland/Games/Spider/Game.cpp:213]:
```suggestion
                    if (m_focused_stack)
                        m_focused_stack->set_focused(false);
                    to_check.set_focused(true);
```

I don't know how exactly but I somehow end up here with `m_focused_stack` being non-null. At least for me this fixes the problem where I can't drag&drop cards.

Comment (gunnarbeutner) [reply Userland/Games/Spider/Game.cpp:213]: FWIW I've been able to reproduce that problem pretty much every game. I just can't tell you what it is I'm doing that triggers it. :)

**Comment by jamierocks** [comment]
I did a little bit of cleanup earlier (implemented victory detection, high scores tracked), and fixed the issue where a draw couldn't complete a stack.

Still need to look into the other issue :)

**Comment by gunnarbeutner** [comment]
Btw, in case you didn't notice: I've added a fix for the drag&drop issue as a review comment. :)

**Comment by jamierocks** [comment]
I did notice, and thank you - I did try to work out what could cause it to be in that state in the first place, but I think applying that patch in the meantime is a wise move.

**Comment by jamierocks** [comment]
I've applied the patch you suggested, thank you :) I've also rebased to a newer commit, though likely out-of-date now haha.

I also made some more polish issues I noticed - showing the correct high score when you change modes for example.

**Comment by AtkinsSJ** [comment]
> 
> Very nice! One thing I've noticed is that sometimes the game won't let me drag&drop a card until after I've moved the target card:
>

I've had this very rarely in Solitaire too, so I think it's an issue in LibCards. My mouse was playing up at the time so I thought it was just that.

**Pull request merged**